proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -0,0 +1,354 @@
+  apply (rule cur_vcpu_valid_lift)
+   apply (clarsimp simp: valid_def)
+   using use_valid[OF _ arch_state] apply (fastforce simp: active_cur_vcpu_of_def)
+  apply (fastforce simp: arch_tcb_of_cur_thread)
+  done
+
+lemma cur_vcpu_valid_lift_cur_thread_update:
+  assumes arch_state: "\<And>P. f \<lbrace>\<lambda>s. P (active_cur_vcpu_of s)\<rbrace>"
+  and arch_tcb_of_cur_thread: "\<And>P. f \<lbrace>\<lambda>s. arch_tcb_at P t s\<rbrace>"
+  shows "f \<lbrace>\<lambda>s. cur_vcpu_valid (s\<lparr>cur_thread := t\<rparr>)\<rbrace>"
+  apply (clarsimp simp: cur_vcpu_valid_def valid_def)
+  using use_valid[OF _ arch_state] use_valid[OF _ arch_tcb_of_cur_thread]
+  by (fastforce simp: active_cur_vcpu_of_def)
+
+lemma as_user_cur_vcpu_valid[wp]:
+  "as_user tptr f \<lbrace>cur_vcpu_valid\<rbrace>"
+  apply (clarsimp simp: as_user_def active_cur_vcpu_of_def)
+  apply (rule cur_vcpu_valid_lift)
+   apply (wpsimp wp: select_f_wp set_object_wp)
+   apply (fastforce simp: active_cur_vcpu_of_def)
+  apply (wpsimp wp: set_object_wp)
+  apply (clarsimp simp: pred_tcb_at_def obj_at_def get_tcb_def)
+  done
+
+lemma machine_state_update_active_cur_vcpu_of[simp]:
+  "P (active_cur_vcpu_of (s\<lparr>machine_state := ms\<rparr>)) = P (active_cur_vcpu_of s)"
+  by (fastforce simp: active_cur_vcpu_of_def)
+
+crunches do_machine_op
+  for cur_vcpu_valid_cur_thread_update[wp]: "\<lambda>s. cur_vcpu_valid (s\<lparr>cur_thread := t\<rparr>)"
+  and cur_vcpu_valid[wp]: "\<lambda>s. cur_vcpu_valid s"
+  (wp: cur_vcpu_valid_lift_cur_thread_update cur_vcpu_valid_lift crunch_wps)
+
+lemma vcpu_save_reg_cur_vcpu_valid[wp]:
+  "vcpu_save_reg vr reg \<lbrace>cur_vcpu_valid\<rbrace>"
+  apply (clarsimp simp: vcpu_save_reg_def vcpu_update_def)
+  apply (rule hoare_seq_ext_skip, wpsimp)
+  apply (wpsimp wp: set_vcpu_wp get_vcpu_wp)
+  apply (clarsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def pred_tcb_at_def obj_at_def)
+  done
+
+lemma vcpu_write_reg_cur_vcpu_valid[wp]:
+  "vcpu_write_reg vr reg val \<lbrace>cur_vcpu_valid\<rbrace>"
+  apply (clarsimp simp: vcpu_write_reg_def vcpu_update_def)
+  apply (wpsimp wp: set_vcpu_wp get_vcpu_wp)
+  apply (clarsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def pred_tcb_at_def obj_at_def)
+  done
+
+lemma save_virt_timer_cur_vcpu_valid[wp]:
+  "save_virt_timer vcpu_timer \<lbrace>cur_vcpu_valid\<rbrace>"
+  apply (clarsimp simp: save_virt_timer_def vcpu_update_def)
+  apply (rule hoare_seq_ext_skip, wpsimp)+
+  apply (wpsimp wp: set_vcpu_wp get_vcpu_wp)
+  apply (clarsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def pred_tcb_at_def obj_at_def)
+  done
+
+lemma vgic_update_cur_vcpu_valid[wp]:
+  "vgic_update vr f \<lbrace>cur_vcpu_valid\<rbrace>"
+  apply (clarsimp simp: vgic_update_def vcpu_update_def)
+  apply (rule hoare_seq_ext_skip, wpsimp)+
+  apply (wpsimp wp: set_vcpu_wp get_vcpu_wp)
+  apply (clarsimp simp: cur_vcpu_valid_def active_cur_vcpu_of_def pred_tcb_at_def obj_at_def)
+  done
```

Comment:
these 4 lemmas look a bit copy-pasta heavy... is there no way to compress or automate this a little better?